File: web/src/IqTest.Domain/Questions/FigureSpec.cs
namespace IqTest.Domain.Questions;

/// <summary>The geometric primitives the matrix items are drawn from.</summary>
public enum ShapeKind
{
    Circle,
    Square,
    Triangle,
    Diamond,
    Hexagon,
    Star,
    Arrow,
}

/// <summary>
/// A declarative description of one cell, or one answer option, of a matrix
/// item. Mirrors the Dart <c>FigureSpec</c> so both platforms draw the same
/// puzzle from the same data.
/// </summary>
/// <param name="Shape">Which primitive to draw.</param>
/// <param name="Count">How many copies the cell holds (1..4).</param>
/// <param name="Filled">Whether the shapes are solid rather than outlined.</param>
/// <param name="RotationQuarters">Clockwise rotation in quarter turns (0..3).</param>
/// <param name="HasDot">Whether each shape carries a centre dot.</param>
public sealed record FigureSpec(
    ShapeKind Shape,
    int Count = 1,
    bool Filled = false,
    int RotationQuarters = 0,
    bool HasDot = false);

File: lib/models/figure_spec.dart
import 'package:flutter/foundation.dart';

/// The geometric primitives the matrix-reasoning items are drawn from.
enum ShapeKind { circle, square, triangle, diamond, hexagon, star, arrow }

/// A declarative description of one cell (or one answer option) in a
/// matrix-reasoning item.
///
/// Items are described by rules over these attributes — shape identity, how
/// many copies are drawn, whether they are filled, their orientation and
/// whether they carry a centre dot — so a puzzle is authored as data and
/// rendered by [FigureView] rather than shipped as an image asset.
@immutable
class FigureSpec {
  const FigureSpec({
    required this.shape,
    this.count = 1,
    this.filled = false,
    this.rotationQuarters = 0,
    this.hasDot = false,
  }) : assert(count >= 1 && count <= 4, 'count must be 1..4'),
       assert(
         rotationQuarters >= 0 && rotationQuarters <= 3,
         'rotationQuarters must be 0..3',
       );

  /// Which primitive to draw.
  final ShapeKind shape;

  /// How many copies of [shape] the cell contains (1..4).
  final int count;

  /// Whether the shapes are solid rather than outlined.
  final bool filled;

  /// Clockwise rotation in quarter turns (0..3).
  final int rotationQuarters;

  /// Whether each shape carries a small dot at its centre.
  final bool hasDot;

  FigureSpec copyWith({
    ShapeKind? shape,
    int? count,
    bool? filled,
    int? rotationQuarters,
    bool? hasDot,
  }) {
    return FigureSpec(
      shape: shape ?? this.shape,
      count: count ?? this.count,
      filled: filled ?? this.filled,
      rotationQuarters: rotationQuarters ?? this.rotationQuarters,
      hasDot: hasDot ?? this.hasDot,
    );
  }

  @override
  bool operator ==(Object other) {
    return other is FigureSpec &&
        other.shape == shape &&
        other.count == count &&
        other.filled == filled &&
        other.rotationQuarters == rotationQuarters &&
        other.hasDot == hasDot;
  }

  @override
  int get hashCode =>
      Object.hash(shape, count, filled, rotationQuarters, hasDot);

  @override
  String toString() =>
      'FigureSpec(${shape.name} x$count, filled: $filled, '
      'rot: $rotationQuarters, dot: $hasDot)';
}
